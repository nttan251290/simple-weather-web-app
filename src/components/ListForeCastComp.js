import React from 'react'
import moment from 'moment-timezone';
import { PropTypes } from 'prop-types'

const ListForeCastComp = props => {
	const { title, list } = props.data

	const WEEKDAYS = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

  return (
    <div className="wrap-forecast">

			<h1 className="title-city">{title}</h1>
			<ul className="list-forecast">

				{ Array.isArray(list) &&	
						list.map(obj => (
							<li className="item" key={obj.id}>
								<div className="content-item">
									<dd className="weatherstate" data-original-title="" title="">
										<div className={`state-icon-sml state-${obj.weather_state_abbr}`}></div>
										<span className="text-weather">{obj.weather_state_name}</span>
									</dd>
									<div className="day">{WEEKDAYS[moment(obj.applicable_date).day()]}</div>
									<div className="min">Min : {parseInt(obj.min_temp)}°C</div>
									<div className="max">Max : {parseInt(obj.max_temp)}°C</div>
								</div>
							</li>
						))
				}
			</ul>

		</div>
  )
}

export default ListForeCastComp;

ListForeCastComp.propTypes = {
	data: PropTypes.object
}
