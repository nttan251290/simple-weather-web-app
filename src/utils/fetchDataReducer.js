const fetchDataReducer = (state, action) => {
  switch (action.type) {
    case "FETCH_INIT":
      return {
        ...state,
        isLoading: true,
        isError: false
      };
    case "FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload.data
      };
    case "FETCH_FAILED":
      return {
        ...state,
        isLoading: false,
        isError: true
      };
    default:
      return state;
  }
};

export default fetchDataReducer