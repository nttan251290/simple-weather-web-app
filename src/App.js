import React, { useState, useEffect } from "react";
import axios from "axios";
import useFetchData from './utils/useFetchData';

import IconLoadingComp from './components/IconLoadingComp';
import ListForeCastComp from './components/ListForeCastComp';

import { URL_LOCATION_SEARCH, URL_LOCATION_FORECAST, URL_LIST_CITIES } from './utils/constant';
import "./styles.css";


function App() {

	const [listCity, setListCity] = useState([]);
	const [city, setCity] = useState('Ho Chi Minh');
	
	const [{ data, isLoading, isError }, setFetchUrl] = useFetchData({
		initialData: {
			data: {}
		},
		urlLocationSearch: URL_LOCATION_SEARCH + city,
		urlLocationForeCast: URL_LOCATION_FORECAST
	});

	useEffect(() => {
		const fetchData = async () => {
			try {
				const { data } = await axios.get(URL_LIST_CITIES);
				setListCity(data.cities);
			} catch (error) {
				console.log(error);
			}
		};

		fetchData();
	}, []);


	return (
		<div className="container">
			<h1 className="title">WEATHER FORECAST</h1>
			<form className="pure-form"
						onSubmit={e => {
							e.preventDefault();
							setFetchUrl(URL_LOCATION_SEARCH + city);
						}}>

				<div className="pure-control-group">
					<input
						className="input-city"
						list="city-list"
						onChange={e => {
							setCity(e.target.value);
						}}
					/>

					<datalist id="city-list">
						{Array.isArray(listCity) && listCity.map((city, index) => <option key={index} value={city}></option>)}
					</datalist>

				</div>
			</form>

			{isError && <label className="error" id="error-msg">Your City you enter is not correct. Please try another city in the list</label>}

			{isLoading && <IconLoadingComp />}

			{!isError && !isLoading && <ListForeCastComp data={data} />}

		</div>
	);
}

export default App;
