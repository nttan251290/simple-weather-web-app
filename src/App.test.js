import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import App from './App';
import IconLoadingComp from './components/IconLoadingComp';
import ListForeCastComp from './components/ListForeCastComp';

import fetchDataReducer from './utils/fetchDataReducer'


describe('IconLoading Component', () => {
	test('IconLoadingComp renders', () => {
		const component = renderer.create(<IconLoadingComp />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('ListForeCastComp Component', () => {
	test('ListForeCastComp renders', () => {
		let data = { title: 'Ho Chi Minh City - Vietnam', list: [{ id: 1, weather_state_abbr: 'hc', weather_state_name: 'Heavy Cloud', applicable_date: '2020-04-29', min_temp: '26', max_temp: 36 }] }

		const component = renderer.create(<ListForeCastComp data={data} />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

describe('App Container Component', () => {
	test('Renders title weather web app', () => {
		const { getByText } = render(<App />);
		const linkElement = getByText('WEATHER FORECAST');
		expect(linkElement).toBeInTheDocument();
	});
});

describe('fetchDataReducer', () => {

	it('should return the INITIAL state', () => {
		expect(fetchDataReducer({}, {})).toEqual({});
	});

})