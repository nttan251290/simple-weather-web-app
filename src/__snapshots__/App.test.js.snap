// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconLoading Component IconLoadingComp renders 1`] = `
<div
  className="sk-chase"
>
  <div
    className="sk-chase-dot"
  />
  <div
    className="sk-chase-dot"
  />
  <div
    className="sk-chase-dot"
  />
  <div
    className="sk-chase-dot"
  />
  <div
    className="sk-chase-dot"
  />
  <div
    className="sk-chase-dot"
  />
</div>
`;

exports[`ListForeCastComp Component ListForeCastComp renders 1`] = `
<div
  className="wrap-forecast"
>
  <h1
    className="title-city"
  >
    Ho Chi Minh City - Vietnam
  </h1>
  <ul
    className="list-forecast"
  >
    <li
      className="item"
    >
      <div
        className="content-item"
      >
        <dd
          className="weatherstate"
          data-original-title=""
          title=""
        >
          <div
            className="state-icon-sml state-hc"
          />
          <span
            className="text-weather"
          >
            Heavy Cloud
          </span>
        </dd>
        <div
          className="day"
        >
          Wednesday
        </div>
        <div
          className="min"
        >
          Min : 
          26
          °C
        </div>
        <div
          className="max"
        >
          Max : 
          36
          °C
        </div>
      </div>
    </li>
  </ul>
</div>
`;
